ONSdigital/rasrm-acceptance-tests
# 937 additional waits and checks

# Motivation and Context
Acceptance tests were failing seemingly randomly. This was an investigation to try and find why.


# What has changed
What was found was : 
1) Chrome driver in headless drops the session after 30 seconds , firefox does not
2)Several tests had finger in the air sleep statements , several of these changed to await for new elements , url changes etc
3) The remaining common failure is a 409 from the survey service which should be fixed by an upcoming release of survey service and tests

This release uses an environment variable called WEBDRIVER . This defaults to chrome , but can (and should) be over ridden to firefox. The dev can switch between firefox and chrome as and when they wish. We can then loop back around after everyone is using firefox and set the default to firefox.

Note gecko writes to geckodriver.log so you may wish to clear that down periodically

# How to test?

1) Install Firefox 
2) Install gecko driver ( e.g brew install geckodriver ) , make sure its on path.
3) export WEBDRIVER=firefox (in .bashrc ?) 

Run acceptance tests . You can switch back to chrome at any time by either deleting WEBDRIVER environment variable or setting it to 'chrome' . 
NB. Remember to set WEBDRIVER in env variables in whatever IDE you use  else you will carry on with chrome

Errors of the type : HTTPError: 409 Client Error: Conflict for url: http://localhost:8080/surveys are expected and should be cleaned up by a later change. But aside from that the tests run a lot more reliably

# Links
https://trello.com/c/QSoSFAU2
